keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Dynamic Resource Allocation uses parameters for resources that are completely
+opaque to core Kubernetes. They get interpreted by the DRA driver's controller
+(for allocating claims) and the DRA drivers' kubelet plugin (for configuring
+the resource on a node). During scheduling of a pod, kube-scheduler and the DRA
+driver controller(s) that handle claims used by the pod communicate through the
+apiserver by updating the PodSchedulingContext object, ultimately leading to
+allocation of all pending claims and the pod being scheduled onto a node.
```

Comment:
So, you make a good point with this comment:

> So basically a configurable mapping of fields in a vendor CRD to the in-tree numeric model type? That adds more flexibility, but also complexity. I don't see how it addresses the "fragility between resources" - instead of one configurable field path, we would have many. Perhaps being explicit about all paths would help. 🤷

I think my initial comment didn't quite hit the right issue. The point is that the interaction between the scheduler and the DRA driver's controller here creates too tight of a coupling that it is hard for things like cluster autoscaler to replicate. What I am trying to get to is a looser coupling. One way we can do that is to separate evaluation of the workload (claims) from reservation/allocation. For instance, we could define a driver-neutral model that represents a node's capacity of different individual resources (and ideally their relationships to one another within various topologies). It would be the responsibility of the kubelet and the  on-node driver agents produce a resource in the api server with this data. This is effectively already what you're doing, just translated into a K8s neutral model rather than captured in driver-specific structures.

Similarly, we would still need (in many cases), a per-driver controller to interpret the driver-specific claims. However instead of the scheduler asking during the scheduling process, we would not even start trying to schedule until each of those driver controllers has contributed to an overall resource request object. In other words, just as we have a driver neutral k8s resource to represent the capacity in the node, we would have a driver-neutral capacity request object in the control plane. The scheduler and other components would deal exclusively in these driver-neutral models.

We may need a way to publish a "schema" describing the specific ways in which allocations can be done out of the model. And we will need a separate resource to represent reservations, so that we can atomically allocate resources across drivers.

Reply:
> One way we can do that is to separate evaluation of the workload (claims) from reservation/allocation.

The numeric parameter approach already does that: evaluation of how to allocate is done in-memory, without any interaction with the apiserver. The once there is a result, the scheduler updates objects to finalize it's decision and get a pod running. This could be extended to schedule multiple pods - it's just not how the scheduler works today.

> This is effectively already what you're doing, just translated into a K8s neutral model rather than captured in driver-specific structures.

The `NodeResourceCapacity` already contains data that is in a format defined by Kubernetes (the capacity type defined for a model). Only the claim parameter object is using a driver-defined type, but that also has a mapping to a Kubernetes type.

> we would have a driver-neutral capacity request object in the control plane

This sounds like you prefer to replace ResourceClaim with something else that is a bit higher level. I'm not sure about that: suppose that one workload needs a GPU and requests it together with some other resources via a single object. One of the use cases is to schedule another "monitoring" pod, potentially with a different life cycle than the workload, which gets access to the same GPU. But that pod then only needs that GPU, not the rest. ResourceClaim is the right level of abstraction for that.

> And we will need a separate resource to represent reservations, so that we can atomically allocate resources across drivers.

That part becomes harder with multiple ResourceClaims per workload. The *decision* can be atomic and doesn't need to worry about concurrent allocation of capacity because the scheduler is the sole owner of the node capacity, but persisting that decision isn't atomic. But the same is true for persisting the pod binding, which also needs to update multiple different objects.